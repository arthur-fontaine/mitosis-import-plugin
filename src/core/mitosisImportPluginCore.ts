import { exec } from "node:child_process";
import { promisify } from "node:util";
import fs from "node:fs";

import type { Target } from "./types/Target";
import path from "node:path";

const execAsync = promisify(exec);

class MitosisImportPluginCore {
	requiresMitosisProcessing(attributes: Record<string, string>): boolean {
		return attributes.mitosis !== undefined;
	}

	getMitosisTarget(attributes: Record<string, string>): Target {
		const target = attributes.mitosis || "auto";
		return target;
	}

	async compileMitosisComponent(
		componentSource: string,
		componentPath: string,
		target: Target,
	): Promise<string> {
		this.#throwIfTargetIsAuto(target);
		const mitosisPath = this.#getMitosisExecutablePath();

		const command = `echo ${JSON.stringify(componentSource)} | ${mitosisPath} compile -t ${target} --path ${componentPath}`;
		try {
			const output = await execAsync(command, { encoding: "utf-8" });
			if (output.stderr) throw new Error(output.stderr);
			return output.stdout;
		} catch (error) {
			const message = error instanceof Error ? error.message : String(error);
			throw new Error(`Error compiling Mitosis component: ${message}`);
		}
	}

	#getMitosisExecutablePath(): string {
		const mitosisCliPackagePath = require.resolve(
			"@builder.io/mitosis-cli/package.json",
		);
		const mitosisCliPackage = JSON.parse(
			fs.readFileSync(mitosisCliPackagePath, "utf-8"),
		);

		const mitosisCliPath = path.join(
			path.dirname(mitosisCliPackagePath),
			mitosisCliPackage.bin.mitosis,
		);

		return mitosisCliPath;
	}

	#throwIfTargetIsAuto(target: Target): void {
		if (target === "auto") {
			throw new Error("Auto target is not yet supported");
		}
	}
}

export const mitosisImportPluginCore = new MitosisImportPluginCore();
