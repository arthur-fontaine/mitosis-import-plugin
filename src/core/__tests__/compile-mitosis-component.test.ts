/// <reference types="vite/client" />

import { describe, expect, test } from "vitest";
import { mitosisImportPluginCore } from "../mitosisImportPluginCore";

describe("compileMitosisComponent", async () => {
	for (const fixtureModulePath in import.meta.glob("./fixtures/*.ts")) {
		const fixtureModule = import.meta.glob("./fixtures/*.ts")[
			fixtureModulePath
		];
		const fixture = (await fixtureModule()) as {
			input: string;
			target: string;
			autoPath?: string;
		};
		const { input, target } = fixture;

		test(`should return the compiled component for fixture: ${fixtureModulePath}`, async () => {
			const componentSource = input.trim();

			const result = await mitosisImportPluginCore.compileMitosisComponent(
				{ source: componentSource, path: "path/to/component.ts" },
				{ source: "", path: "" },
				target,
			);

			expect(result).toBeDefined();
			expect(result).toMatchSnapshot();
		});

		if (fixture.autoPath) {
			test(`should return the component compiled to the guessed target for fixture: ${fixtureModulePath}`, async () => {
				const componentSource = input.trim();

				const result = await mitosisImportPluginCore.compileMitosisComponent(
					{ source: componentSource, path: "path/to/component.ts" },
					// biome-ignore lint/style/noNonNullAssertion: we checked that autoPath is defined above
					{ source: "", path: fixture.autoPath! },
					"auto",
				);

				expect(result).toBeDefined();
				expect(result).toMatchSnapshot();
			});
		}
	}
});
