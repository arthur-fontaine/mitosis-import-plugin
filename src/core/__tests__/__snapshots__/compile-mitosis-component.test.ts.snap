// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`compileMitosisComponent > should return the compiled component for fixture: ./fixtures/get-started-svelte.ts 1`] = `
"<script>
  import onChange from "on-change";

  name = "Steve";
</script>

<div>
  <input class="input" bind:value={state.name} />
  Hello! I can run natively in React, Vue, Svelte, Qwik, and many more frameworks!
</div>

<style>
  .input {
    color: red;
  }
</style>
"
`;

exports[`compileMitosisComponent > should return the compiled component for fixture: ./fixtures/warnings-in-stdout.ts 1`] = `
"/** @jsx h */
import { h, Fragment } from "preact";

function SftpDownsyncIcon(props) {
  return <span>SFTP</span>;
}

export default SftpDownsyncIcon;

"
`;
